hphp/runtime/vm/jit/vasm-reuseimm.cpp

```diff
@@ -0,0 +1,212 @@
+  Vunit& unit;
+  ImmStateAry& immStateAry;
+};
+
+bool isMultiword(int64_t imm) {
+  switch (arch()) {
+  case Arch::X64:
+  case Arch::PPC64:
+    break;
+  case Arch::ARM:
+    uint64_t val = std::abs(imm);
+    if (val > (1 << 16)) return true;
+    break;
+  }
+  return false;
+}
+
+// candidate around +/- uimm12
+bool reuseCandidate(Env& env, int64_t p, Vreg& reg, int& delta) {
+  delta = 0;
+  for (int i = 0; i < env.immStateAry.size; i++) {
+    ImmState elem = env.immStateAry[i];
+    if (!elem.base.isValid()) continue;
+    int64_t q = elem.val.q();
+    if ((p >= q) && (p < (q + 4095))) {
+      delta = safe_cast<int>(p - q);
+      reg = elem.base;
+      return true;
+    }
+    if ((p < q) && (q < (p + 4095))) {
+      delta = -safe_cast<int>(q - p);
+      reg = elem.base;
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename Inst>
+void reuseImmq(Env& env, const Inst& inst, Vlabel b, size_t i) { 
+  env.immStateAry[i] = ImmState{};
+  return;
+}
+
+// purge table across any instruction with RegSet or control flow change
+#define Y(vasm_opc) \
+void reuseImmq(Env& env, const vasm_opc& c, Vlabel b, size_t i) { \
+  env.immStateAry.clear(); \
+  return; \
+} 
+
+Y(fallthru)
+Y(call)
+Y(callarray)
+Y(callfaststub)
+Y(callm)
+Y(callr)
+Y(calls)
+Y(callstub)
+Y(calltc)
+Y(contenter)
+Y(jcc)
+Y(jcci)
+Y(jmpi)
+Y(jmpm)
+Y(jmpr)
+Y(landingpad)
+Y(leavetc)
+Y(nothrow)
+Y(phpret)
+Y(ret)
+Y(syncpoint)
+Y(ud2)
+Y(unwind)
+
+#undef Y
+
+
+template<typename ReuseImm>
+void reuseimm_impl(Vunit& unit, Vlabel b, size_t i, ReuseImm reuse) {
+  vmodify(unit, b, i, [&] (Vout& v) { reuse(v); return 1; });
+}
+
+void reuseImmq(Env& env, const ldimmq& ld, Vlabel b, size_t i) {
+  if (isMultiword(ld.s.q())) {
+    int off;
+    Vreg base;
+    if (reuseCandidate(env, ld.s.q(), base, off)) {
+     if (off >= 0 ) {
+       reuseimm_impl(env.unit, b, i, [&] (Vout& v) {
+         v << addqi{off, base, ld.d, v.makeReg()};
+       });
+     } else {
+       reuseimm_impl(env.unit, b, i, [&] (Vout& v) {
+         v << subqi{-off, base, ld.d, v.makeReg()};
+       });
+      }
+    return;
+    }
+  }
+  env.immStateAry[i] = ImmState{ld.s, ld.d};
+}
+
+void reuseImmq(Env& env, Vlabel b, size_t i) {
+  assertx(i <= env.unit.blocks[b].code.size());
+  auto const& inst = env.unit.blocks[b].code[i];
+
+  switch (inst.op) {
+#define O(name, ...)                      \
+    case Vinstr::name:                    \
+      reuseImmq(env, inst.name##_, b, i); \
+      return;
+
+  VASM_OPCODES
+#undef O
+  }
+  not_reached();
+}
+
+}
+
+// Opportunistically reuse immediate q values
+void reuseImmq(Vunit& unit) {
+  assertx(check(unit));
+  auto& blocks = unit.blocks;
+
+  if (RuntimeOption::EvalJitLdimmqSpan <= 0) {
+    return;
+  }
+
+  ImmStateAry immStateAry;
+  Env env { unit, immStateAry };
+
+  auto const labels = sortBlocks(unit);
+  env.immStateAry.size =
+    std::min<int>(RuntimeOption::EvalJitLdimmqSpan, ImmStateAry::max_history);
```

Comment:
Why not just let the RuntimeOption dictate everything?  You can just use a `std::vector` instead of a C-style array.